src/online2bin/online2-net-nnet3-latgen-faster.cc

```diff
@@ -0,0 +1,398 @@
+
+                server.Write(msg + '\r');
+              }
+
+              next_check += check_update;
+            }
+
+          } else {
+            feature_pipeline.InputFinished();
+
+            decoder.FinalizeDecoding();
+
+            if (decoder.NumFramesDecoded() > 0) {
+              Lattice lat;
+              decoder.GetBestPath(true, &lat);
+
+              std::string msg = LatticeToString(lat, word_syms);
+
+              server.Write(msg + "\n");
+            }
+
+            server.disconnect();
+            break;
+
+          }
+
+          if (decoder.EndpointDetected(endpoint_opts)) {
+            server.Write("\n");
+            feature_pipeline.GetAdaptationState(&adaptation_state);
+            break;
+          }
+        }
+      }
+    }
+  } catch (const std::exception &e) {
+    std::cerr << e.what();
+    return -1;
+  }
+} // main()
+
+
+namespace kaldi {
+TcpServer::TcpServer() {
+  server_desc_ = -1;
+  client_desc_ = -1;
+  samp_buf_ = NULL;
+  buf_len_ = 0;
+}
+
+bool TcpServer::Listen(int32 port) {
+  h_addr_.sin_addr.s_addr = INADDR_ANY;
+  h_addr_.sin_port = htons(port);
+  h_addr_.sin_family = AF_INET;
+
+  server_desc_ = socket(AF_INET, SOCK_STREAM, 0);
+
+  if (server_desc_ == -1) {
+    KALDI_ERR << "Cannot create TCP socket!";
+    return false;
+  }
+
+  int32 flag = 1;
+  int32 len = sizeof(int32);
+  if (setsockopt(server_desc_, SOL_SOCKET, SO_REUSEADDR, &flag, len) == -1) {
+    KALDI_ERR << "Cannot set socket options!\n";
+    return false;
+  }
+
+  if (bind(server_desc_, (struct sockaddr *) &h_addr_, sizeof(h_addr_)) == -1) {
+    KALDI_ERR << "Cannot bind to port: " << port << " (is it taken?)";
+    return false;
+  }
+
+  if (listen(server_desc_, 1) == -1) {
+    KALDI_ERR << "Cannot listen on port!";
+    return false;
+  }
+
+  KALDI_LOG << "TcpServer: Listening on port: " << port << std::endl;
+
+  return true;
+
+}
+
+TcpServer::~TcpServer() {
+  disconnect();
+  if (server_desc_ != -1)
+    close(server_desc_);
+  delete[] samp_buf_;
+}
+
+int32 TcpServer::Accept() {
+  KALDI_LOG << "Waiting for client..." << std::endl;
+
+  socklen_t len;
+
+  len = sizeof(struct sockaddr);
+  client_desc_ = accept(server_desc_, (struct sockaddr *) &h_addr_, &len);
+
+  struct sockaddr_storage addr;
+  char ipstr[20];
+
+  len = sizeof addr;
+  getpeername(client_desc_, (struct sockaddr *) &addr, &len);
+
+  struct sockaddr_in *s = (struct sockaddr_in *) &addr;
+  inet_ntop(AF_INET, &s->sin_addr, ipstr, sizeof ipstr);
+
+  KALDI_LOG << "Accepted connection from: " << ipstr << std::endl;
+
+  return client_desc_;
+}
+
+bool TcpServer::ReadChunk(size_t len) {
+  if (buf_len_ != len) {
+    buf_len_ = len;
+    delete[] samp_buf_;
+    samp_buf_ = new int16[len];
+  }
+
+  size_t to_read = len;
+  has_read_ = 0;
+  while (to_read > 0) {
+    ssize_t ret = read(client_desc_, static_cast<void *>(samp_buf_ + has_read_), to_read * sizeof(int16));
+    if (ret <= 0)
+      break;
```

Comment:
This loop won't end unless an error occurs or the client shuts down its end of connection (TCP half-shutdown).

Reply:
Ok, I added a select w/timeout for reading. I'm pretty sure it's not really necessary for writing - it's unlikely we will overflow the output buffer, but I can do that as well if you think it's necessary. I tested the program and it's stable as usual.

Reply:
I don't think that's necessary.